src/lib.rs

```diff
@@ -293,6 +293,77 @@ pub fn resourceprep(s: &str) -> Result<Cow<'_, str>, Error> {
     Ok(Cow::Owned(normalized))
 }
 
+fn x520_mapped_to_nothing(c: char) -> bool {
+    if c.is_control() {
+        return true;
+    }
+    match c {
+        '\u{00AD}' | '\u{1806}' | '\u{034F}' | '\u{180B}'..='\u{180D}' |
+        '\u{FE00}'..='\u{FE0F}' | '\u{FFFC}' | '\u{200B}' => true,
+        _ => false,
+    }
+}
+
+/// Prepares a string according to the procedures described in Section 7 of
+/// [ITU-T Recommendation X.520 (2019)](https://www.itu.int/rec/T-REC-X.520-201910-I/en).
+///
+/// Note that this function does _not_ remove leading, trailing, or inner
+/// spaces as described in Section 7.6, because the characters needing removal
+/// will vary across the matching rules and ASN.1 syntaxes used.
+pub fn x520prep(s: &str, case_fold: bool) -> Result<Cow<'_, str>, Error> {
+    if s.chars().all(|c| matches!(c, ' '..='~') && (!case_fold || c.is_ascii_lowercase())) {
+        return Ok(Cow::Borrowed(s));
+    }
+
+    // 1. Transcode
+    // Already done because &str is enforced to be Unicode.
+
+    // 2. Map
+    let mapped = s.chars()
+        .filter(|&c| !x520_mapped_to_nothing(c))
+        .map(|c| if c.is_whitespace() { ' ' } else { c });
+
+    // 3. Normalize
+    let normalized = if case_fold {
+        mapped
+            .flat_map(tables::case_fold_for_nfkc)
+            .collect::<String>()
+    } else {
+        mapped.collect::<String>()
+    };
+
+    // 4. Prohibit
+    let prohibited = normalized.chars().find(|&c| tables::unassigned_code_point(c)
+        || tables::private_use(c)
+        || tables::non_character_code_point(c)
+        || tables::surrogate_code(c)
+        || c == '\u{FFFD}' // REPLACEMENT CHARACTER
+    );
+    if let Some(c) = prohibited {
+        return Err(Error(ErrorCause::ProhibitedCharacter(c)));
+    }
+    // From ITU-T Recommendation X.520, Section 7.4:
+    // "The first code point of a string is prohibited from being a combining character."
+    let first_char = s.chars().next();
+    if let Some(c) = first_char {
+        if tables::unicode_mark_category(c) {
+            // I do think this ought to be considered a different error, but adding
+            // another enum variant would be a breaking change, so this is "good"
+            return Err(Error(ErrorCause::ProhibitedCharacter(first_char.unwrap())));
+        }
+    }
```

Comment:
Empty strings are also prohibited.

Reply:
What should I do about this one? There isn't an error code for this, and I don't want to break compatibility. Personally, I feel like since the string is returned to the caller anyway, they can check if it is empty. But I'll add a new `ErrorCause` if you'd prefer that.

Reply:
To add to that, my "let the caller deal with it" excuse was used to avoid implementing the requirements of 7.6 as well, since the treatment of remaining whitespace at that point is conditional based on the matching rule used (X.500 directories have a concept of "matching rules" which describe how two values are compared).

Reply:
Speaking of which, I did consider implementing functions that extend `x520_prep` by doing the insignificant space removal described in section 7.6 (e.g. `x520_numstr_prep`), but I don't want to bloat your code base with X.500 directory-specific code (although I don't think it would be much code to just remove space characters).

Reply:
How would that break back compat?

Reply:
~~Adding another enum variant means that existing code that is currently performing an exhaustive match on all variants will no longer handle all variants. The only way existing could _wouldn't_ be broken is if it uses the "all other cases" `_` branch.~~

I just noticed that `ErrorCode` is not published, although `Error` is. Would you _like_ to see a new enum variant?

Reply:
Yeah I think that'd make sense for this case.